Intro (#51)
infinum/eightshift-docs: website/docs/basics/autowiring.md

---
id: autowiring
title: Autowiring
sidebar_label: Autowiring
---

[![docs-source](https://img.shields.io/badge/source-eigthshift--libs-blue?style=for-the-badge&logo=php&labelColor=2a2a2a)](https://github.com/infinum/eightshift-libs)

What is autowiring?

*Autowiring is an exotic word that represents something very simple: the container's ability to create and inject dependencies automatically.*

The quote was used from [PHP-DI](https://php-di.org/doc/autowiring.html) library that we use.

> In a nutshell, we use a dependency injection container to load all your service classes. With DI container, we also use the PSR-4 coding standard that allows us to predict how your file/folder structure will look like in correlation to the namespace. With that in mind, you don't have to worry about when your class is required, and when it loads, our autowiring takes care of everything. You need to follow a few coding standards.

Reviewer: ```suggestion
In a nutshell, we use a dependency injection container to load all your service classes. Alongside the DI container, we are also using the PSR-4 coding standard for autoloading. That allows us to predict how your file/folder structure will look like in correlation to the namespace. With that in mind, you don't have to worry about requiring your classes. Our autoloading takes care of everything. All you need to do is follow a few coding standards so that you can automatically inject other classes as dependencies.
```

> Autowiring runs inside the `Main` service class as an extended class in our `AbstractMain` class. You can install the main class using this command:

Reviewer 2: ```suggestion
Autowiring runs inside the `Main` service class as an extended class of our `AbstractMain` class. You can install the main class using this command:
```

`wp boilerplate create_main`

## How it works?

Here is a quick overview of how this works:

- You create a new class manually or using the WP-CLI command.
- PSR-4 would detect your new class if you followed the PSR-4 naming standard.

> - New class is added to the `classmap array` inside the `vendor` folder.

Reviewer 2: ```suggestion
- New class is added to the `classmap` array inside the `vendor` folder.
```

- Autowiring class reads the new class and checks if you have any DI classes used in your new class.
- It just works.

As you saw from the overview steps, autowiring checks your class and loads DI classes accordingly.

### What if my class doesn't have anything inside a constructor method?

If your class **doesn't have** anything defined in the constructor method, autowiring will require your class.
This just means that you are not using DI.

### What if my class does have a primitive parameter inside a constructor method?

> If your class ** has** a primitive parameter defined in the constructor method, autowiring will check what type it is and load it to your new class.

Reviewer: If you have a primitive you need to define it at the point of usage. If that point of usage is services array, then yes, you'll need to define them there.

I mean, since we are trying to have just one point of origin for our dependencies (the service array) it's kinda valid in all cases I think 🤔

> If your class ** has** a primitive parameter defined in the constructor method, autowiring will check what type it is and load it to your new class.

Author: 👍

> 

Reviewer 3: I'd add the following here.

### What if my class has an **interface parameter** inside the constructor method?

This will automatically be resolved (same as with class parameters) if you follow 1 simple rule: **Variable name in your constructor method needs to match the class name (which implements the interface) in camelCase.**

For example, let's say you have a `SteeringWheel` class.

```php
class SteeringWheel implements CarPartInterface
```

...and you want to have this as a dependency in `Car` class. Your `Car` class has the following constructor:

```php
public function __construct(CarPartInterface $steeringWheel) {
  $this->steeringWheel = $steeringWheel;
}
```

Since the injected variable name (`$steeringWheel`) is the `camelCase` version of the class name (`SteeringWheel`) which implements `CarPartInterface`, autowiring will know to inject the correct class and everything will work fine.

> ### What if my class does have another class as a parameter inside a constructor method?

Reviewer 3: As discussed, I'd separate these sections like so:

### Injecting primitives

See comment above

### Injecting classes (make it shorter and sum up to "this works BUT you shouldn't do this, see why")

This works out of the box. **But** you shouldn't really do this, see why(link to section "Why you should inject Interfaces and not Classes".

### Injecting interfaces

Works but you need to match the var name (see comment above for this section's contents).

### Why you should inject Interfaces and not Classes (move everything that's currently in this section here)

Explanation about why coding against interfaces is better than coding against implementations.

In a nutshell, we use a dependency injection container to load all your service classes. With DI container, we also use the PSR-4 coding standard that allows us to predict how your file/folder structure will look like in correlation to the namespace. With that in mind, you don't have to worry about when your class is required, and when it loads, our autowiring takes care of everything. You need to follow a few coding standards.

Autowiring runs inside the `Main` service class as an extended class in our `AbstractMain` class. You can install the main class using this command:

`wp boilerplate create_main`

## How it works?

Here is a quick overview of how this works:

- You create a new class manually or using the WP-CLI command.
- PSR-4 would detect your new class if you followed the PSR-4 naming standard.
- New class is added to the `classmap array` inside the `vendor` folder.
- Autowiring class reads the new class and checks if you have any DI classes used in your new class.
- It just works.

As you saw from the overview steps, autowiring checks your class and loads DI classes accordingly.

### What if my class doesn't have anything inside a constructor method?

> If your class **doesn't have** anything defined in the constructor method, autowiring will require your class.

Reviewer: ```suggestion
If your class **doesn't have** anything defined in the constructor method, autoloading will require your class.
```

> This just means that you are not using DI.

Reviewer: ```suggestion
This just means that you are not using dependency injection (as you have nothing to inject).
```

> ### What if my class does have a primitive parameter inside a constructor method?

Reviewer: ```suggestion
### What if my class has a **primitive parameter** (`string`, `int`, etc.) inside a constructor method?
```

Adding on top of Grga's suggestion

If your class ** has** a primitive parameter defined in the constructor method, autowiring will check what type it is and load it to your new class.

### What if my class does have another class as a parameter inside a constructor method?

> Well, this is when things become interesting. A good coding practice is that your class should never depend on the same class because you have tightly coupled your class to another class. This way, you can't simply reuse your class, test it, or copy it to another project. 

Reviewer: ```suggestion
This is when things become interesting. A good coding practice is that your class should never depend on the concrete class implementation because you have tightly coupled your class to another class. This makes it hard to test and your code becomes hard to modify. Imagine that you have put a concrete implementation as a dependency, only to get feedback from the client that you need to change that implementation for a completely different one. Making the changes means that you'll need to track all the places in your codebase where you have used some functionality from this class, and change it completely.
```

**You should always code against interfaces and not implementation.**

> We can't stress this enough because as your project grows, so will your headaches. Also, when you start testing your code, that is when your hair will begin to fall off. So read [Uncle Bob Martins - Clean Code](http://cleancoder.com/products) book and save your self a lot of sleepless nights.

Reviewer: ```suggestion
We can't stress this enough because as your project grows, so will your headaches. Also, when you start testing your code, that is when your hair will begin to fall off. We recommend reading Uncle Bob Martin's [Clean Code](http://cleancoder.com/products) book. That will save you a lot of sleepless nights, and you'll learn tons of tips and tricks for writing clean code.
```

> Let's set a scene. You have a created class `DocumentsRoute` class and your class need some functionality from the `QueryDocuments` class.

Reviewer: ```suggestion
Let's set a scene. You have created a `DocumentsRoute` class and it needs some functionality from the `QueryDocuments` class.
```

> What are your steps:

Reviewer: ```suggestion
Steps you need to follow are:
```

> - You should create an interface with the method you need in the `DocumentsRoute` class, let's call it `QueryDocumentsInterface.class` and put it inside the `src/Query/Documents` folder.

Reviewer: ```suggestion
- You should create an interface with all the methods you need in the `DocumentsRoute` class, let's call it `QueryDocumentsInterface` and we'll put it inside the `src/Query/Documents` folder.
```

> - Your `QueryDocuments` class should implement your new `QueryDocumentsInterface` class and have all the methods from the interface defined.

Reviewer: ```suggestion
- Your `QueryDocuments` class should implement your new `QueryDocumentsInterface` class and have all the methods from the interface implemented.
```

> - In your `DocumentsRoute` class provide the `QueryDocumentsInterface` interface as a parameter inside the `__construct` method.

Reviewer: ```suggestion
- In your `DocumentsRoute` class provide the `QueryDocumentsInterface` interface as a parameter inside the `__construct` method with the name of the argument $queryDocuments.
```

> - Provide some additional markup to use this `QueryDocumentsInterface` in your `DocumentsRoute` class.

Reviewer: ```suggestion
- You can now use the methods defined in the `QueryDocuments` class.
```

> Sounds confusing, right?

Reviewer: I'd remove this. No need to remind the reader that something may be confusing 😅

> Let's us provide you with some code examples:

Reviewer: ```suggestion
Let's provide you with some code examples:
```

**Structure:**
- src
	- Query
		- Documents
			- QueryDocumentsInterface.class
			- QueryDocuments.class
	- Rest
		- Routes
			- DocumentsRoute.class

> **src/Query/Documents/QueryDocumentsInterface.class**

Reviewer: ```suggestion
**src/Query/Documents/QueryDocumentsInterface**
```

> ```js

Reviewer: ```suggestion
```php
```

/**

>  * The file that is an Query for Documents class interface.

Reviewer: ```suggestion
 * The file that holds a QueryDocuments interface.
```

 *
 * @package ProjectNamespace\Query\Documents;
 */

declare(strict_types=1);

namespace ProjectNamespace\Query\Documents;

/**

>  * Interface for a QueryDocumentsInterface

Reviewer: ```suggestion
 * Interface for querying documents
```

 */
interface QueryDocumentsInterface
{

  /**
   * Get structured documents data for documents section.
   *
   * @param int $id Id for data.
   *
   * @return string
   */
  public function getDocumentsData(int $id = 0): string;
}
```

> **src/Query/Documents/QueryDocuments.class**

Reviewer: ```suggestion
**src/Query/Documents/QueryDocuments.php**
```

> ```js

Reviewer: ```suggestion
```php
```

/**

>  * The file that is an Query of Documents class.

Reviewer: ```suggestion
 * The file for querying all the documents
```

 *
 * @package ProjectNamespace\Query\Documents
 */

declare(strict_types=1);

namespace ProjectNamespace\Query\Documents;

use ProjectNamespaceVendor\EightshiftLibs\Services\ServiceInterface;

/**
 * QueryDocuments class.
 */
class QueryDocuments implements ServiceInterface, QueryDocumentsInterface
{
    /**
   * Get structured documents data for documents section.
   *
   * @param int $id Id for data.
   *
   * @return string
   */
  public function getDocumentsData(int $id = 0): string
  {

>     ... here goes the implementation of getDocumentsData method.

Reviewer: ```suggestion
    // here goes the implementation of getDocumentsData method.
```

  }
}
```

> **src/Rest/Routes/DocumentsRoute.class**
> ```js

Reviewer: ```suggestion
**src/Rest/Routes/DocumentsRoute.php**
```php
```

/**
 * The class register route for DocumentsRoute endpoint
 *
 * @package ProjectNamespace\Rest\Routes
 */

declare(strict_types=1);

namespace ProjectNamespace\Rest\Routes;

use ProjectNamespace\Query\Documents\QueryDocumentsInterface;
use ProjectNamespaceVendor\EightshiftLibs\Rest\Routes\AbstractRoute;
use ProjectNamespaceVendor\EightshiftLibs\Rest\CallableRouteInterface;

/**
 * Class DocumentsRoute
 */
class DocumentsRoute extends AbstractRoute implements CallableRouteInterface
{

  /**
   * Instance variable of documents data.
   *
   * @var QueryDocumentsInterface
   */

>   protected $documentsData;
> 
>   /**
>    * Create a new instance.
>    *
>    * @param QueryDocumentsInterface $queryDocuments Inject documentsData which holds data for documents.
>    */
>   public function __construct(QueryDocumentsInterface $queryDocuments)
>   {
>     $this->documentsData = $queryDocuments;

Reviewer: ```suggestion
  protected $queryDocuments;

  /**
   * Create a new instance.
   *
   * @param QueryDocumentsInterface $queryDocuments Inject documentsData which holds data for documents.
   */
  public function __construct(QueryDocumentsInterface $queryDocuments)
  {
    $this->queryDocuments = $queryDocuments;
```

It's a standard practice to name the instance variable the same name as the dependency.

  }

>   ...here goes the rest of the DocumentsRoute implementation.

Reviewer: ```suggestion
  // here goes the rest of the DocumentsRoute implementation.
```

}
```

> An this is it, autowiring knows how to resolve this setup, and you are all set.

Reviewer: ```suggestion
And that's it. Autowiring knows how to resolve the dependencies, you didn't tightly couple the code, and you're all set.
```

### What if I have to mock or do some weird stuff in my DI?

> Sometimes automatization can save you, but in the end, eventually, you need something custom, we got you covered.

Reviewer: ```suggestion
Automatization can save you, but in the end, if you need something custom, we got you covered.
```

We can think of these scenarios:
- You have to write tests and manually provide mocked classes.

> - You need to do some weird stuff in your DI.

Reviewer: ```suggestion
```

What is weird stuff? I never did anything weird in my DIC 😂 

We can remove anything that is overly vague 😉

- Your classes have some custom structure, and autowiring can't resolve it.

In those cases, you can manually provide your DI container with the implementation using the `getServiceClasses` method inside the `src>Main>Main` class.

Just provide the method and add your custom implementation like this:

> `src>Main>Main.class`

Reviewer: ```suggestion
`src>Main>Main.php`
```

> ```js

Reviewer: ```suggestion
```php
```

  /**
   * Get the list of services to register.
   *
   * A list of classes which contain hooks.
   *
   * @return array<class-string, string|string[]> Array of fully qualified service class names.
   */
  protected function getServiceClasses(): array
  {
    return [

      // If you are using a class as a DI.

>       ProjectNamespace\Rest\Routes\DocumentsRoute => [ProjectNamespace\Query\Documents\QueryDocuments],

Reviewer: ```suggestion
      ProjectNamespace\Rest\Routes\DocumentsRoute::class => [ProjectNamespace\Query\Documents\QueryDocuments::class],
```

      // If you just want to include a simple class with no DI.

>       ProjectNamespace\CoolFolder\CoolClass.class,

Reviewer: ```suggestion
      ProjectNamespace\CoolFolder\CoolClass::class,
```

    ];
  }
```

> ## Sum it up

Reviewer: ```suggestion
## To sum it up
```

> We understand that this is a long chapter but take your time and try to understand it because it can make you a better developer, your project better, and your clients happier. We tried to explain this as simply as possible.

Reviewer: ```suggestion
We understand that this is a long chapter but take your time and try to understand it because it can make you a better developer, your project better, and your clients happier. We tried to explain this as simply as possible.
```

Again, we should avoid making these claims. Using DIC doesn't automatically make you a better developer (oh how I wish it did 😄 ). Creating a good architecture and making your code testable will help you out in the long run (and that's the aim of all this). You can still use DIC and make shitty architecture that will bite you in the ass 😂 

I'd go with something along the lines of:

This chapter may be a bit long, but we wanted to make sure you understand our motivation behind it. We often work on large projects, with lots of third-party integrations, and having a good project structure and architecture makes the projects more durable. It makes them less prone to bugs (if you write tests), makes the implementation of new features, and removing obsolete features a lot easier.

Do you need to use it? Not really. Especially on really simple projects. Should you? We think you'll benefit from it in the long run.
